src/alfasim_sdk/result_reader/aggregator.py

```diff
@@ -0,0 +1,1344 @@
+    time_sets_unit = attr.ib(validator=attr.validators.instance_of(str))
+    time_steps_boundaries = attr.ib(validator=attr.validators.instance_of(tuple))
+    time_set_info = attr.ib(validator=attr.validators.instance_of(dict))
+    result_directory = attr.ib(validator=attr.validators.instance_of(Path))
+    app_version_info = attr.ib(validator=attr.validators.instance_of(dict))
+
+
+@attr.s(slots=True, hash=False)
+class _MergedMetadataWithStatistics:
+    """
+    :ivar app_version_info:
+        Map "base time steps" to the application version used in the simulation.
+
+    :ivar profile_key_to_metadata:
+        Map output keys (profile id and property id) to metadata entries.
+
+    :ivar trend_key_to_metadata:
+        Map output keys (trend id and property id) to metadata entries.
+
+    :ivar time_set_keys:
+        List of sourced time set keys (tuples of relevant base time sets).
+    """
+
+    app_version_info = attr.ib(validator=attr.validators.instance_of(dict))
+    profile_key_to_metadata = attr.ib(validator=attr.validators.instance_of(dict))
+    trend_key_to_metadata = attr.ib(validator=attr.validators.instance_of(dict))
+    time_set_keys = attr.ib(validator=attr.validators.instance_of(list))
+
+
+def load_result_files(result_directory: Path) -> Dict[int, h5py.File]:
+    """
+    Return a dict with the result files.
+
+    Note that once the container dict is collected the files originally returned are closed.
+    """
+
+    def open_result_file(filename: Path) -> h5py.File:
+        # The lib h5py will fail to open the file for some exotic paths,
+        # but will work if the cwd are such paths.
+        directory = filename.parent
+        old_directory = os.getcwd()
+        try:
+            os.chdir(directory)
+            try:
+                return h5py.File(filename.name, "r", libver="latest", swmr=True)
+            except OSError as os_error:
+                swmr_message = (
+                    "Unable to open file (file is not already open for SWMR writing)"
+                )
+                if str(os_error) == swmr_message:
+                    return h5py.File(filename.name, "r", libver="latest", swmr=False)
+                raise
+
+        except PermissionError:
+            raise PermissionError(
+                f"Could not access folder {filename}.\n"
+                f'If you are using your local "Downloads" directory, consider changing the project file\n'
+                f"to somewhere else as some services try to synchronize that directory and interfere with\n"
+                f"the simulation.\n"
+                "\n"
+                f'1 - Click in "Save as".\n'
+                f"2 - Select a different path, outside Download folder.\n"
+                f"3 - Run simulation again.\n"
+            )
+        finally:
+            os.chdir(old_directory)
+
+    # When a new result file is created its metadata contents are not complete, and the file
```

Comment:
A safer approach would be to actually write the result files to a temporary file name (in the same directory), using an extension which wouldn't be globbed here, and only when the file is complete, you move the file to the final name using `os.replace`.

You can then simplify this code because now there's no special handling that needs to be done.

Or am I missing something?

Reply:
On Windows when moving the file the writer process need to close the file.
In the time window between "move" and "reopen the file in swmr write mode" a reader could try to open the file in swmr reading mode, this will fail (a swmr writter must open the file first) and the open in normal mode fallback will be used, causing inconsistencies when the writer process reopen the file.
I actually don't remember if the "open for swmr write" will fail or not (open to read when no writer is presente raises).

I welcome suggestions to improve the comment =)

Reply:
Not sure, that was just a normal comment... but are you aware that `os.replace` is atomic (contrary to `os.rename` which is not)?

Reply:
Why can't you `put hdf in swmr mode` in the first case before closing? Sorry I don't remember the swmr details.

Reply:
Or you can only `put hdf in swmr mode` for the file if it is in its final filename?

Reply:
The swmr works by the writer opening the file first, when closing the file is no longer open.

Reply:
Ahh got it. 👍 Thanks for the explanation.